### PR TITLE
Add confirm modal for pin, unpin message

### PR DESCRIPTION
**Pull request by giachuvan**

### CODE REVIEW

File: crates/mezon-ui/src/chat/pinned_popover.rs [hunk 1]
```diff
@@ -116,15 +108,20 @@ impl PinnedPopoverPanel {
         }
     }
 
+    fn refresh_name_rows(&mut self, cx: &mut Context<Self>) {
+        let count = self.list_state.item_count();
+        if count > 0 {
+            self.list_state.reset(count);
```
Comment (huylexuan44) [inline crates/mezon-ui/src/chat/pinned_popover.rs]: reset() nulls logical_scroll_top — list snaps to top on every member store notify. Use measure_all() or just cx.notify().

File: crates/mezon-store/src/badge.rs [hunk 2]
```diff
@@ -308,9 +308,18 @@ impl BadgeService {
                         }
                     }
                     if !from_me && is_new_message {
-                        ClanList::global(cx).update(cx, |cls, cx| {
-                            cls.set_has_unread_message(clan_id, cx);
-                        });
+                        let code = MessageCode::from_raw(m.code);
+                        if !matches!(
+                            code,
+                            MessageCode::CreatePin
```
Comment (huylexuan44) [inline crates/mezon-store/src/badge.rs]: note_channel_message above is unguarded — CreatePin still marks channel unread and sync_clan_after_read relights the nub.

File: crates/mezon-store/src/pinned.rs [hunk 3]
```diff
@@ -267,34 +308,137 @@ impl PinnedMessagesStore {
...
+                } else if let Some(channel_id) = channel_id_opt
+                    && let Some(url) =
+                        resolve_avatar_url(user_id, ProfileContext::Direct(channel_id), cx)
+                            .filter(|url| !url.is_empty())
+                {
+                    avatar = url;
+                }
+            }
+            let content_plain = msg.content.clone();
+            let content_wire = serde_json::json!({ "t": content_plain }).to_string();
```
Comment (huylexuan44) [inline crates/mezon-store/src/pinned.rs]: React sends the full content JSON; {"t": text} drops mk/hg/ej/embeds for remote clients until they reload.

File: crates/mezon-store/src/pinned.rs [hunk 4]
```diff
@@ -267,34 +308,137 @@ impl PinnedMessagesStore {
...
+                avatar_proxied: avatar_proxied.into(),
+                content: content_plain,
+                create_time,
+            },
+        );
+        self.set_pin_badge(&channel_id_str, cx);
+        cx.notify();
+
         let api = self.api.clone();
+        let attachment = "[]".to_string();
```
Comment (huylexuan44) [inline crates/mezon-store/src/pinned.rs]: Attachments hardcoded to []. React sends the filtered attachment array — pinned image previews break on other clients.

File: crates/mezon-ui/src/chat/message/pin_confirm_modals.rs [hunk 5]
```diff
@@ -0,0 +1,622 @@
...
+        if PinnedMessagesStore::global(cx)
+            .read(cx)
+            .is_pinned(&message_id.to_string())
+        {
+            return;
+        }
+
+        let (message, is_dm, mut clan_id, channel_id) = {
+            let messages = MessagesStore::global(cx).read(cx);
+            let Some(msg) = messages.messages().iter().find(|m| m.id == message_id) else {
```
Comment (huylexuan44) [inline crates/mezon-ui/src/chat/message/pin_confirm_modals.rs]: Silent return when the message isn't in the active list: no modal, no feedback. Use find_channel_message fallback + warn.

File: crates/mezon-ui/src/chat/pinned_popover.rs [hunk 6]
```diff
@@ -307,111 +320,101 @@ fn render_body(
...
     theme: &Theme,
     locale: &str,
     popover_handle: PopoverMenuHandle<PinnedPopoverPanel>,
     avatar_cache: Entity<LruImageCache>,
+    cx: &App,
 ) -> gpui::AnyElement {
     let tokens = &theme.tokens;
     let group_name = SharedString::from(format!("pin-card-{index}"));
     let sender_color = Hsla::from(gpui::rgb(DEFAULT_DISPLAY_NAME_COLOR));
+    let sender_label = resolve_pin_sender_label_with_message(
```
Comment (huylexuan44) [inline crates/mezon-ui/src/chat/pinned_popover.rs]: Sender label and avatar now resolved per frame inside the list closure. Precompute into PinCardVm as before.

File: crates/mezon-ui/src/chat/message/pin_confirm_modals.rs [hunk 7]
```diff
@@ -0,0 +1,622 @@
...
+        )
+    }
+}
+
+impl ConfirmUnpinMessageModal {
+    pub fn open(
+        pin_id: SharedString,
+        message_id: SharedString,
+        sender_label: SharedString,
+        _avatar_src: Option<SharedString>,
```
Comment (huylexuan44) [inline crates/mezon-ui/src/chat/message/pin_confirm_modals.rs]: _avatar_src, _avatar_fallback and _content are unused — preview() re-derives them. Drop these three params.

File: crates/mezon-ui/src/chat/message/pin_confirm_modals.rs [hunk 8]
```diff
@@ -0,0 +1,622 @@
...
+        .p(px(8.))
+        .rounded(px(4.))
+        .bg(tokens.bg_item_theme_hover)
+        .child(avatar)
+        .child(
+            v_flex().flex_1().min_w_0().gap_1().child(name_row).child(
+                div()
+                    .text_sm()
+                    .text_color(tokens.text_theme_message)
+                    .child(preview.content.clone()),
```
Comment (huylexuan44) [inline crates/mezon-ui/src/chat/message/pin_confirm_modals.rs]: No max_h or line clamp — a long pinned message makes an unbounded-height modal. Content also renders as raw text.

File: crates/mezon-ui/src/chat/channel_header.rs [hunk 9]
```diff
@@ -633,6 +633,10 @@ impl ChannelHeader {
             {
                 let is_open = handle.is_deployed();
                 let menu_handle = handle.clone();
+                let show_badge = PinnedMessagesStore::global(cx)
+                    .read(cx)
+                    .active_has_pin_badge();
+                let badge_color = gpui::rgb(0xda_37_3c);
```
Comment (huylexuan44) [inline crates/mezon-ui/src/chat/channel_header.rs]: Hardcoded 0xda373c. Use a theme token for React parity.

File: crates/mezon-store/src/pinned.rs [hunk 10]
```diff
@@ -225,6 +230,41 @@ impl PinnedMessagesStore {
         .detach();
     }
 
+    pub fn is_pinned(&self, message_id: &str) -> bool {
+        self.messages.iter().any(|m| m.message_id == message_id)
+    }
+
+    pub fn has_pin_badge(&self, channel_id: &str) -> bool {
```
Comment (huylexuan44) [inline crates/mezon-store/src/pinned.rs]: has_pin_badge has no callers. Wire it up or delete it.

File: crates/mezon-ui/src/chat/inbox/panel.rs [hunk 11]
```diff
@@ -1107,5 +1107,5 @@ pub fn clan_has_inbox_badge(clan_id: &str, cx: &App) -> bool {
         .clans
         .iter()
         .find(|c| c.id == clan_id)
-        .is_some_and(|c| c.badge_count > 0 || c.has_unread)
+        .is_some_and(|c| c.badge_count > 0)
```
Comment (huylexuan44) [inline crates/mezon-ui/src/chat/inbox/panel.rs:1108]: Unrelated to pin modals — split into its own commit or explain the has_unread removal in the PR description.

File: crates/mezon-store/src/message.rs [hunk 12]
```diff
@@ -1172,6 +1175,32 @@ fn strip_marker(s: &str, marker: &str) -> String {
     trimmed.to_string()
 }
 
+fn strip_code_fence(s: &str) -> (Option<String>, String) {
```
Comment (huylexuan44) [inline crates/mezon-store/src/message.rs]: Regression: the newline is stripped before the language check, so an un-tagged code fence loses its first line. Also hits the main chat list.

File: crates/mezon-ui/src/chat/pinned_popover.rs [hunk 13]
```diff
@@ -516,6 +509,705 @@ fn pin_card(
...
+}
+
+fn flush_pin_text_buf(buf: &mut String, out: &mut Vec<MessageSpan>) {
+    if buf.is_empty() {
+        return;
+    }
+    out.push(MessageSpan::Text(std::mem::take(buf).into()));
+}
+
+fn normalize_pin_code_block(raw: &str) -> SharedString {
```
Comment (huylexuan44) [inline crates/mezon-ui/src/chat/pinned_popover.rs]: Duplicates strip_code_fence and runs again at 594 and 821, so a 3-line code block renders as 1 line. Strip once in the store.

File: crates/mezon-ui/src/chat/pinned_popover.rs [hunk 14]
```diff
@@ -516,6 +509,705 @@ fn pin_card(
         .into_any_element()
 }
 
+fn render_pin_body(
```
Comment (huylexuan44) [inline crates/mezon-ui/src/chat/pinned_popover.rs:585]: render.md forbids markdown parsing in render(); measure_all plus remeasure() re-parses every pinned row on member-store notify.

File: crates/mezon-ui/src/chat/pinned_popover.rs [hunk 15]
```diff
@@ -273,11 +299,17 @@ fn render_body(
...
+            .flex()
+            .flex_col()
             .pl(px(LIST_PAD_X))
             .pr(px(LIST_PAD_X))
-            .py(px(8.))
+            .py(px(LIST_PAD_Y))
             .child(
-                list(list_state.clone(), move |ix, _window, _cx| {
+                list(list_state.clone(), move |ix, _window, cx| {
+                    let store = PinnedMessagesStore::global(cx).read(cx);
```
Comment (huylexuan44) [inline crates/mezon-ui/src/chat/pinned_popover.rs]: Row mixes a live store read with a snapshot vm; if the store mutates mid-frame pin[ix] and vm[ix] are different messages. Put the body in PinCardVm.

File: crates/mezon-ui/src/chat/message/pin_confirm_modals.rs [hunk 16]
```diff
@@ -0,0 +1,627 @@
...
+    }
+}
+
+fn member_subscriptions(cx: &mut Context<ConfirmPinMessageModal>) -> Vec<Subscription> {
+    vec![
+        cx.observe(&ClanMembersStore::global(cx), |_, _, cx| cx.notify()),
+        cx.observe(&UsersByUserStore::global(cx), |_, _, cx| cx.notify()),
+        cx.observe(&AccountStore::global(cx), |_, _, cx| cx.notify()),
+        cx.observe(&DirectMessageStore::global(cx), |_, _, cx| cx.notify()),
+        cx.observe(&MessagesStore::global(cx), |_, _, cx| cx.notify()),
```
Comment (huylexuan44) [inline crates/mezon-ui/src/chat/message/pin_confirm_modals.rs]: invalidation.md: broad observe on MessagesStore fires on every message, and it is dead here since message is a snapshot clone.

File: crates/mezon-store/src/pinned.rs [hunk 17]
```diff
@@ -267,34 +314,168 @@ impl PinnedMessagesStore {
...
+        cx.notify();
+
         let api = self.api.clone();
         cx.spawn(async move |_this, _cx| {
             if let Err(e) = api
-                .create_pin_message(message_id, channel_id, clan_id)
+                .create_pin_message(message_id_i64, channel_id_i64, clan_id_i64)
                 .await
             {
                 tracing::error!("create_pin_message failed: {e}");
```
Comment (huylexuan44) [inline crates/mezon-store/src/pinned.rs:462]: Optimistic row is never rolled back when create_pin_message fails. unpin() refreshes on error; do the same here.

File: crates/mezon-store/src/badge.rs [hunk 18]
```diff
@@ -307,7 +315,7 @@ impl BadgeService {
                             });
                         }
                     }
-                    if !from_me && is_new_message {
+                    if !from_me && is_new_message && !skip_unread_activity {
```
Comment (huylexuan44) [inline crates/mezon-store/src/badge.rs:318]: skip_unread_activity is not applied to increment_clan_badge at 325, so a CreatePin mention can bump the clan badge with nothing to clear it.

File: crates/mezon-ui/src/chat/message/pin_confirm_modals.rs [hunk 19]
```diff
@@ -0,0 +1,627 @@
...
+            v_flex().flex_1().min_w_0().gap_1().child(name_row).child(
+                div()
+                    .id("pin-confirm-preview")
+                    .w_full()
+                    .min_w_0()
+                    .max_h(px(320.))
+                    .overflow_y_scroll()
+                    .text_sm()
+                    .text_color(tokens.text_theme_message)
+                    .child(preview.content.clone()),
```
Comment (huylexuan44) [inline crates/mezon-ui/src/chat/message/pin_confirm_modals.rs]: Preview renders plain text; React shows a full MessageWithUser. Attachments, code blocks and mentions are invisible here.

File: crates/mezon-ui/src/chat/message/pin_confirm_modals.rs [hunk 20]
```diff
@@ -0,0 +1,627 @@
...
+
+        let (is_dm, mut clan_id, channel_id) = {
+            let messages = MessagesStore::global(cx).read(cx);
+            (
+                messages.is_dm(),
+                messages.active_clan_id(),
+                messages.active_channel_id(),
+            )
+        };
+        let Some(message) = find_channel_message(message_id, channel_id, cx) else {
```
Comment (huylexuan44) [inline crates/mezon-ui/src/chat/message/pin_confirm_modals.rs]: Silent no-op when the message is not in MessagesStore; before this PR pin only needed the id. Add a toast or an id-only fallback.

File: crates/mezon-store/src/pinned.rs [hunk 21]
```diff
@@ -267,34 +314,168 @@ impl PinnedMessagesStore {
...
+        } else {
+            return;
+        };
+
+        let cfg = AppConfig::try_global(cx);
+        let avatar_proxied = cfg
+            .map(|c| c.avatar_proxy(&avatar_url))
+            .unwrap_or_else(|| avatar_url.clone());
+        let body = enrich_pin_body(&content_wire, pin_attachments, cfg);
+        let attachment = if body.attachments.is_empty() {
```
Comment (huylexuan44) [inline crates/mezon-store/src/pinned.rs:411]: React filters attachments by isValidUrl before broadcasting; an in-flight upload can be sent with an empty url.

File: crates/mezon-ui/src/chat/pinned_popover.rs [hunk 22]
```diff
@@ -506,7 +499,7 @@ fn pin_card(
         .child(avatar)
         .child(
             v_flex()
-                .flex_1()
+                .w(relative(0.85))
```
Comment (huylexuan44) [inline crates/mezon-ui/src/chat/pinned_popover.rs]: Magic ratio replacing flex_1; breaks if avatar size or padding changes. Use flex_1().min_w_0().

File: crates/mezon-ui/src/chat/pinned_popover.rs [hunk 23]
```diff
@@ -516,6 +509,705 @@ fn pin_card(
...
+    }
+    let rest = &line[hashes..];
+    let body = rest.trim_start_matches([' ', '\t']);
+    if body.len() == rest.len() || body.is_empty() {
+        return None;
+    }
+    Some((hashes as u8, body))
+}
+
+fn soft_break_hints(text: &str) -> String {
```
Comment (huylexuan44) [inline crates/mezon-ui/src/chat/pinned_popover.rs]: Injects U+200B into every run including code blocks, so copied code carries invisible chars. Also allocates per run per frame.

File: crates/mezon-ui/src/chat/pinned_popover.rs [hunk 24]
```diff
@@ -516,6 +509,705 @@ fn pin_card(
...
+        .max_w_full()
+        .overflow_hidden()
+        .text_sm()
+        .line_height(rems(1.25))
+        .text_color(color)
+        .child(soft_break_hints(text))
+        .into_any_element()
+}
+
+fn pin_heading_size(level: u8) -> gpui::Pixels {
```
Comment (huylexuan44) [inline crates/mezon-ui/src/chat/pinned_popover.rs]: Byte-identical copy of content.rs heading_size/heading_line_height. Reuse instead of forking.

File: crates/mezon-store/src/pinned.rs [hunk 25]
```diff
@@ -326,18 +507,190 @@ impl PinnedMessagesStore {
...
 }
 
+fn pin_content_wire(msg: &Message) -> String {
+    if let Some(raw) = msg.raw_content.as_deref().filter(|raw| !raw.is_empty()) {
+        return raw.to_string();
+    }
+    rebuild_pin_content_json(msg)
+}
+
+fn rebuild_pin_content_json(msg: &Message) -> String {
```
Comment (huylexuan44) [inline crates/mezon-store/src/pinned.rs:530]: mk/hg/ej/lk offsets come from marker-stripped span text while t keeps the markers, so the ranges drift.
